journal/src/main/scala/akka/persistence/spanner/internal/SpannerObjectInteractions.scala

```
@@ -0,0 +1,260 @@
+            case newSeqNr @ ("new_sequence_nr", TypeCode.INT64) => wrapNameValue(newSeqNr, seqNr)
+            case other => throw new MatchError(other)
+          }), SqlUpdateTypes.mapValues(Type(_)).toMap)
+        )
+      )(session)
+      .map(response => {
+        val updated = response.resultSets.head.stats.head.rowCount.rowCountExact.head
+        if (updated != 1L)
+          throw new IllegalStateException(
+            s"Update failed: object for key [$key] could not be updated to sequence number [$seqNr]"
+          )
+      })
+
+  // TODO maybe return timestamp?
+  def getObject(key: String): Future[Result] =
+    spannerGrpcClient
+      .withSession(
+        session =>
+          spannerGrpcClient.client.read(
+            ReadRequest(
+              session.session.name,
+              transaction = None,
+              settings.objectTable,
+              index = "",
+              Seq(Objects.Value, Objects.SerId, Objects.SerManifest, Objects.SeqNr).map(_._1),
+              wrapKey(key),
+              limit = 1
+            )
+          )
+      )
+      .map(
+        resultSet =>
+          resultSet.rows.headOption match {
+            case Some(ListValue(Seq(value, serId, serManifest, seqNr), _)) =>
+              Result(
+                ByteString(value.getStringValue).decodeBase64,
+                serId.getStringValue.toLong,
+                serManifest.getStringValue,
+                seqNr.getStringValue.toLong
+              )
+            case None =>
+              throw ObjectNotFound(key)
+          }
+      )
+
+  def deleteObject(key: String): Future[Unit] =
+    spannerGrpcClient.withSession(session => {
+      spannerGrpcClient.write(
+        Seq(
+          Mutation(
+            Mutation.Operation.Delete(
+              Mutation.Delete(
+                settings.objectTable,
+                wrapKey(key)
+              )
+            )
+          )
+        )
+      )(session)
```

what happens when trying to delete a key that doesn't exist? should that be a failed Future or a silent success?

---

Is it, though? A failure that goes unnoticed may cause more problems than some noise...

---

If deletes are only supposed to be performed by the single-writer entity it should know that the object exists or not. In that case it is useful to detect that it was an invalid delete. Otherwise idempotency could be a reason for silently accepting delete of already deleted.